0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,82 @@
+
+An advantage of PSK is that the pre-shared secret key only needs to be shared once. So long as both parties possess this key and are listening for transfers, they can send payments between one another.
+
+A disadvantage of PSK is that it is repudiable. Knowledge of a payment's fulfillment cannot be used as cryptographic proof that the receiver completed the payment, because both sender and receiver know how to generate the fulfillment with their pre-shared secret key.
+
+## Flow
+
+0. The pre-shared secret key is shared out of band.
+0. The sender creates an inner (PRIVATE) PSK request, with information on the expiry and any additional data enclosed within.
+0. The sender creates a random Key-Id value, and calculates an HMAC of the Key-Id using the pre-shared secret key as the key. This HMAC will be referred to as the payment key.
+0. The sender encrypts the inner request with the payment key.
+0. The sender creates an outer (PUBLIC) header, and attaches the Key-Id and HMAC function they used as headers. The encrypted inner request is attached as binary data.
+0. The sender quotes to the receiver and sends a transfer with the outer header in the data field of the ILP packet. The fulfillment of this payment is the HMAC of the ILP packet using the payment key as the key.
+0. The receiver gets notified of in incoming prepared transfer with an ILP packet.
+0. The receiver generates the payment key from the outer request contained in the data field of the ILP packet, and uses it to decrypt the inner request.
+0. _(Optional)_ The receiver calls a backend to review the payment.
+0. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the payment key as the key.
+
+## Data Format
+
+The PSK details are divided into two request objects in the style of [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). The data of the outer request is the encrypted inner request.
+
+Each of the requests starts with `PSK/1.0` and then `PUBLIC` or `PRIVATE`.  The requests are made of a list of headers, followed by data. For simplicity, duplicate headers are not allowed. 
+
+The headers of the outer request are visible to all parties transmitting the request, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the inner request.
+
+### Outer Request
+
+The outer request is formatted as below:
+
+```http
+PSK/1.0 PUBLIC
+Key-Id: fpwpAhlN588
+Key-Algorithm: HMAC-SHA-256
+Unsafe-Header: data_everyone_can_see
+
+...
+```
+
+Everything in this request is readable by all connectors between the sender and receiver. PSK payments use an [HMAC](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code) of the PSK details as a fulfillment, in order to prove that the contents are unmodified.  The pre-shared secret key is the key used in the HMAC.
```

Comment:
Above you called the HMAC key the "payment key". We should use names consistently